proposals/0000-tyop-quantification-order.rst

```diff
@@ -0,0 +1,137 @@
+  ordering. Thus, the variables in ``f`` will be ``b``, then ``a``.
+
+The problem this proposal aims to fix is that the actual implementation of
+implicit quantification in GHC does not adhere to the left-to-right rule in two
+cases:
+
+1. Type operators ``a `tyop` b``.
+   Expected quantification order (left-to-right): ``forall a tyop b.``;
+   actual quantification order: ``forall tyop a b.``.
+
+2. Multiplicity annotations ``a %p -> b``.
+   Expected quantification order (left-to-right): ``forall a p b.``;
+   actual quantification order: ``forall a b p.``.
+
+It is rather clear that this is nothing more than a bug. It is also the case
+that fixing this bug constitutes a breaking change. The fix is rather
+straightforward and available at `GHC MR !11036 <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11036/>`_.
+
+The question is whether we want to ship the fix right away or introduce a
+compatibility warning first and wait a few releases.
+
+We argue that there is no need for the warning, as there is practically no code
+in the wild that relies on the current buggy behavior. See the impact
+assessment in the "Backward Compatibility" section.
+
+Proposed Change Specification
+-----------------------------
+* Use left-to-right order in implicit quantification of type variables,
+  including variables that occur as type operators or multiplicity annotations.
+
+* Effect the fix immediately, without a compatibility warning or a wait period.
+
+
+Examples
+--------
+
+1. Type variable used as a type operator:
+   ::
+
+     f :: a `op` b
+
+     vta_old = f @(,) @Int @Bool     -- before the change
+     vta_new = f @Int @(,) @Bool     -- after the change
+
+
+2. Type variable used as a multiplicity annotation:
+   ::
+
+     h :: a %m -> b
+     h = ...
+
+     vta_old = h @Int @Bool @One     -- before the change
+     vta_new = h @Int @One @Bool     -- after the change
+
+Effect and Interactions
+-----------------------
+The fix makes the implementation adhere to the specification laid out in the
+paper and the User's Guide.
+
+Costs and Drawbacks
+-------------------
+No known costs or drawbacks, other than potential breakage in exotic cases.
+
+Backward Compatibility
+----------------------
+Technically, this is a breaking change, as demonstrated by the examples in the
+"Examples" section. At the same time, we expect that practically no code in the
+wild relies on the current buggy behavior.
+
+To estimate the actual impact, we patched GHC 9.6 to use the proposed implicit
+quantification rules and compiled 3337 packages (`full list <https://gist.github.com/int-index/e9d305961d5540f1519492c956e3e65c>`_).
+There have been no build failures, confirming our hypothesis that the breakage
+is observed only in artificial examples.
+
+Alternatives
+------------
+1. Keep the current (incorrect) quantification order. This means adding a
+   special case to the specification instead of changing the implementation.
```

Comment:
Alternative 1.5: do only the fix for `%m`, because it only affects experimental code, and the other has been here for a long time and we would bite that particular bullet.

Reply:
Added it as a third alternative.